src/audio/smart_amp_aw/aw_smart_amp_wapper.c

```diff
@@ -0,0 +1,758 @@
+	}
+
+	smart_amp_sktune_free(sktune);
+
+	return 0;
+}
+
+
+int smart_amp_check_audio_fmt(int sample_rate, int ch_num)
+{
+
+	if (sample_rate != 4800) {
+		return -1;
+	}
+
+	if (ch_num != 2) {
+		return -1;
+	}
+
+	return 0;
+}
+
+int smart_amp_flush(sktune_t *sktune, struct comp_dev *dev)
+{
+	memset(sktune->frame_in.data_ptr, 0,
+	       SMART_AMP_FF_BUF_DB_SZ * sizeof(int32_t));
+	memset(sktune->frame_out.data_ptr, 0,
+	       SMART_AMP_FF_BUF_DB_SZ * sizeof(int32_t));
+	memset(sktune->frame_iv.data_ptr, 0,
+	       SMART_AMP_FB_BUF_DB_SZ * sizeof(int32_t));
+
+
+	comp_dbg(dev, "[Awinic] Reset (handle:%p)", sktune);
+
+	return 0;
+}
+
+
+static int smart_amp_get_buffer(char *buf, uint32_t frames,
+				const struct audio_stream __sparse_cache *stream, uint32_t num_ch)
+{
+	int idx;
+	uint32_t sample_num = frames * num_ch;
+	int byte_num = 0;
+	int16_t *input16, *output16;
+	int32_t *input32, *output32;
+
+	input16 = audio_stream_get_rptr(stream);
+	input32 = audio_stream_get_rptr(stream);
+	output16 = (int16_t *)buf;
+	output32 = (int32_t *)buf;
+
+	switch (audio_stream_get_frm_fmt(stream)) {
+	case SOF_IPC_FRAME_S16_LE:
+		for (idx = 0 ; idx < sample_num ; idx++) {
+				input16 = audio_stream_read_frag_s16(stream, idx);
+				output16[idx] = *input16;
+		}
+		byte_num =  sample_num * sizeof(int16_t);
+		break;
+	case SOF_IPC_FRAME_S24_4LE:
+	case SOF_IPC_FRAME_S32_LE:
+		for (idx = 0 ; idx < sample_num ; idx++) {
+				input32 = audio_stream_read_frag_s32(stream, idx);
+				output32[idx] = *input32;
+		}
+		byte_num =  sample_num * sizeof(int32_t);
+		break;
+	default:
+		return -1;
+	}
+	return byte_num;
+}
+
+
+int smart_amp_fb_data_prepare(sktune_t *       sktune, struct comp_dev *dev,
+				const struct audio_stream __sparse_cache *source,
+			     uint32_t frames)
+
+{
+	int num_ch, ret;
+
+	if (frames == 0) {
+		comp_warn(dev, "[Awinic] feedback frame size zero warning.");
+		return 0;
+	}
+
+	if (frames > SMART_AMP_FB_BUF_DB_SZ) {
+		comp_err(dev, "[Awinic] feedback frame size overflow  : %d", frames);
+		return -1;
+	}
+
+	num_ch = MIN(audio_stream_get_channels(source), SMART_AMP_FB_MAX_CH_NUM);
+
+	ret = smart_amp_get_buffer(sktune->frame_iv.data_ptr, frames, source, num_ch);
+	if (ret < 0) {
+		comp_err(dev, "[Awinic] get IV buf failed");
+		sktune->frame_iv.actual_data_len = 0;
+		return -1;
+	}
+	
+	sktune->frame_iv.actual_data_len = (uint32_t)ret;
+
+
+	return 0;
+}
+
+
+int smart_amp_ff_data_prepare(sktune_t *       sktune, struct comp_dev *dev,
+				const struct audio_stream __sparse_cache *source,
+			     uint32_t frames)
+{
+	int num_ch, ret;
+	
+	if (frames == 0) {
+		comp_warn(dev, "[Awinic] ff frame size zero warning.");
+		return 0;
+	}
+
+	if (frames > SMART_AMP_FF_BUF_DB_SZ) {
+		comp_err(dev, "[Awinic] ff frame size overflow  : %d", frames);
+		return -1;
+	}
+
+	num_ch = MIN(audio_stream_get_channels(source), SMART_AMP_FB_MAX_CH_NUM);
+
+	ret = smart_amp_get_buffer(sktune->frame_in.data_ptr, frames, source, num_ch);
```

Comment:
if the only thing that varies here is the data_ptr then i feel like this and the above function can be generalized together.

Reply:
Once the algorithm is stable, we'll optimize the code